File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
def solution(user_id, banned_id):
    answer = set()

    def is_matched(user, banned):
        if len(user) != len(banned):
            return False
        
        # b가 *이 아니고, u랑 b가 다르면 불량 아님
        for u, b in zip(user, banned):
            if b != '*' and u != b:
                return False
        return True

    def dfs(idx, visited):
        # 끝까지 다 왔음
        if idx == len(banned_id):
            # 중복 방지 -> set으로 만들어서 넣어주기
            # set은 안되고 frozenset이라고 해야한대요...
            answer.add(frozenset(visited))
            return
        
        for user in user_id:
            # 방문 확인
            if user in visited:
                continue
            # 불량 사용자 잡히는지 확인
            if not is_matched(user, banned_id[idx]):
                continue

            visited.append(user)
            dfs(idx + 1, visited)
            visited.pop()

    dfs(0, [])
    return len(answer)

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;

class Solution {

    Set<Set<String>> answer = new HashSet<>();

    public int solution(String[] user_id, String[] banned_id) {

        dfs(0, new HashSet<>(), user_id, banned_id);

        return answer.size();
    }

    boolean isMatched(String user, String banned) {
        if (user.length() != banned.length()) {
            return false;
        }

        // b가 *이 아니고, u랑 b가 다르면 불량 아님
        for (int i = 0; i < user.length(); i++) {
            char u = user.charAt(i);
            char b = banned.charAt(i);

            if (b != '*' && u != b) {
                return false;
            }
        }

        return true;
    }

    void dfs(int idx, Set<String> visited, String[] user_id, String[] banned_id) {
        // 끝까지 다 왔음
        if (idx == banned_id.length) {
            // 중복 방지 -> set으로 만들어서 넣어주기
            answer.add(new HashSet<>(visited));
            return;
        }

        for (String user : user_id) {
            // 방문 확인
            if (visited.contains(user)) {
                continue;
            }

            // 불량 사용자 잡히는지 확인
            if (!isMatched(user, banned_id[idx])) {
                continue;
            }

            visited.add(user);
            dfs(idx + 1, visited, user_id, banned_id);
            visited.remove(user);
        }
    }
}